### ISSUE TITLE
sphinx_internal_class_name greatly increases search time

### ISSUE THREAD

**Issue by jrust** [pat/thinking-sphinx#568]
We rolled out TS3 in production today and immediately noticed the CPU on our searchd box spiked to over 100% and search page time about doubled.  All the sphinx queries were taking quite a bit longer.  The culprit turned out to be removing the line that adds `@sphinx_internal_class_name SomeClass` to every search query.  It appears that sphinx has to work a lot harder to do that field search and in our case it having the class name appended to every search was unnecessary since the indices were specified.

I'm not sure why sphinx_internal_class_name was added to TS3 (support for polymorphism?), but it seems like it'd be good to only add that to the search if it's needed.

**Comment by cweilemann** [comment]
@pat I hate to bring back an old ghost, but if we are _not_ using STI, can we safely remove `sphinx_internal_class_name` from our sphinx.conf file? Or do we need it as an `attr` for any reason?

**Comment by pat** [comment]
Hi Chris

I'd expect `sphinx_internal_class_name` to not appear in SQL-backed indices unless the model's inheritance column is present (and thus, inheritance of some sort is in play).

However, I'm not sure if this behaviour is matched with real-time indices. Can you confirm what type of indices you're using and which version of Thinking Sphinx?

**Comment by cweilemann** [comment]
> I'd expect sphinx_internal_class_name to not appear in SQL-backed indices unless the model's inheritance column is present (and thus, inheritance of some sort is in play).

That's what I'd expect too... Though we haven't removed it because we don't want to break anything.

> However, I'm not sure if this behaviour is matched with real-time indices. Can you confirm what type of indices you're using and which version of Thinking Sphinx?

ThinkingSphinx version is 3.1.4. I'm not a Sphinx expert... I know we're using the real-time, and I'm pretty sure on-disk as well, but I'd have to check.

**Comment by pat** [comment]
So I've pushed a fix for this to the `develop` branch (48921d6b28ef8fcd59be460404af90937473e55e) - the automatic removal of the `sphinx_internal_class_name` field wasn't happening for real-time indices. So, to use this change, you can modify your Gemfile to include the following:

```ruby
gem "thinking-sphinx", "~> 3.3.0",
  :git    => "git://github.com/pat/thinking-sphinx.git",
  :branch => "develop",
  :ref    => "48921d6b28"
```

Granted, this means you'll need to step through upgrading from 3.1.4 to 3.3.0 (the release notes listed in the README should cover all the relevant changes).

The only commit that may cause headaches that is more recent than 3.3.0 is 839d480f72dc811d57be396846b86bed564b1c72 - I'd hope that'll actually be fine, but to be on the safe side (and for future upgrading friendliness), any filters should be cast to their appropriate types.
